mm2src/lp_ordermatch/lp_bot.rs

```diff
@@ -94,25 +195,100 @@ impl Default for Provider {
 #[derive(Default)]
 pub struct TradingBotContext {
     trading_bot_states: AsyncMutex<TradingBotState>,
-    trading_bot_cfg: AsyncMutex<SimpleMakerBotRegistry>,
-    price_url: AsyncMutex<String>,
+}
+
+impl TradingBotContext {
+    async fn get_refresh_rate(&self) -> f64 {
+        let state = self.trading_bot_states.lock().await;
+        if let TradingBotState::Running(running_state) = &*state {
+            return running_state.bot_refresh_rate;
+        }
+        BOT_DEFAULT_REFRESH_RATE
+    }
+}
+
+#[derive(Clone)]
+pub struct ArcTradingBotContext(Arc<TradingBotContext>);
+
+impl Deref for ArcTradingBotContext {
+    type Target = TradingBotContext;
+    fn deref(&self) -> &TradingBotContext { &*self.0 }
+}
+
+#[allow(clippy::single_match)]
+impl TradingBotContext {
+    async fn on_trading_bot_event(&self, ctx: &MmArc, trading_bot_event: &TradingBotEvent) {
+        let msg_format = format!("{}", trading_bot_event);
+        info!("{}", msg_format);
+        let message_service_ctx = MessageServiceContext::from_ctx(ctx).unwrap();
+        let message_service = message_service_ctx.message_service.lock().await;
```

Comment:
I expected that dispatcher will take the responsibility to forward an event to a specific handler like `message_service`. Could you please clarify why you use dispatcher in 
https://github.com/KomodoPlatform/atomicDEX-API/blob/d0216261e85488cda936821921d6f4749a6d8945/mm2src/lp_ordermatch/simple_market_maker.rs#L284-L287
but `message_service` here?

Reply:
> I expected that dispatcher would take the responsibility to forward an event to a specific handler like `message_service`. Could you please clarify why you use dispatcher in
> 
> https://github.com/KomodoPlatform/atomicDEX-API/blob/d0216261e85488cda936821921d6f4749a6d8945/mm2src/lp_ordermatch/simple_market_maker.rs#L284-L287
> 
> 
> but `message_service` here?

Currently, the bot here is used as the dispatcher to send types of events: When the bot starts and stop. On receipt of events, the service message is used to deliver notification messages. The idea here is that there will be more event listeners (Websocket, this type of event can be listened to and treated differently depending on the listeners).

Here the internal bot listener is only used to deliver notification messages. (Telegram, WhatsApp, Messenger and other kinds of msg services that accept text).

The idea is that different listeners can listen to the same types of events and process them in different ways.

Reply:
I get the idea now, thanks for the clarification! :slightly_smiling_face: